packages/dev/core/src/Rendering/IBLShadows/iblShadowsVoxelRenderer.ts

```diff
@@ -0,0 +1,623 @@
+     */
+    constructor(scene: Scene, iblShadowsRenderPipeline: IblShadowsRenderPipeline, resolutionExp: number = 6, triPlanarVoxelization: boolean = true) {
+        this._scene = scene;
+        this._engine = scene.getEngine() as Engine;
+        this._triPlanarVoxelization = triPlanarVoxelization;
+        this._renderPipeline = iblShadowsRenderPipeline;
+        if (!this._engine.getCaps().drawBuffersExtension) {
+            Logger.Error("Can't do voxel rendering without the draw buffers extension.");
+        }
+
+        if (this._engine instanceof WebGPUEngine) {
+            this._maxDrawBuffers = 8; // TODO - get this from the WebGPU engine?
+        } else {
+            this._maxDrawBuffers = (this._engine as Engine)._gl.getParameter((this._engine as Engine)._gl.MAX_DRAW_BUFFERS);
+        }
+
+        this._copyMipEffectRenderer = new EffectRenderer(this._engine);
+        this._copyMipEffectWrapper = new EffectWrapper({
+            engine: this._engine,
+            fragmentShader: "pass",
+            useShaderStore: true,
+            uniformNames: [],
+            samplerNames: ["textureSampler"],
+        });
+
+        this.voxelResolutionExp = resolutionExp;
+    }
+
+    private _generateMipMaps() {
+        const iterations = Math.ceil(Math.log2(this._voxelResolution));
+        for (let i = 1; i < iterations + 1; i++) {
+            this._generateMipMap(i);
+        }
+    }
+
+    private _generateMipMap(lodLevel: number) {
+        // Generate a mip map for the given level by triggering the render of the procedural mip texture.
+        const mipTarget = this._mipArray[lodLevel - 1];
+        if (!mipTarget) {
+            return;
+        }
+        mipTarget.setTexture("srcMip", lodLevel === 1 ? this.getVoxelGrid() : this._mipArray[lodLevel - 2]);
+        mipTarget.render();
+    }
+
+    private _copyMipMaps() {
+        const iterations = Math.ceil(Math.log2(this._voxelResolution));
+        for (let i = 1; i < iterations + 1; i++) {
+            this._copyMipMap(i);
+        }
+    }
+
+    private _copyMipMap(lodLevel: number) {
+        // Now, copy this mip into the mip chain of the voxel grid.
+        // TODO - this currently isn't working. "textureSampler" isn't being properly set to mipTarget.
+        const mipTarget = this._mipArray[lodLevel - 1];
+        if (!mipTarget) {
+            return;
+        }
+        const rt = (this.getVoxelGrid() as any)._rtWrapper;
+        if (rt) {
+            this._copyMipEffectRenderer.saveStates();
+            const bindSize = mipTarget.getSize().width;
+
+            // Render to each layer of the voxel grid.
+            for (let layer = 0; layer < bindSize; layer++) {
+                this._engine.bindFramebuffer(rt, 0, bindSize, bindSize, true, lodLevel, layer);
+                this._copyMipEffectRenderer.applyEffectWrapper(this._copyMipEffectWrapper);
+                this._copyMipEffectWrapper.effect.setTexture("textureSampler", mipTarget);

```

Comment:
@sebavan , here I'm trying to get the `pass` shader to copy my mip (`mipTarget`) into the mip chain of the voxel grid. When viewing with Spector.js, it appears that `textureSampler` is set to a completely different texture.

Reply:
yup answered on the forum, I ll review again once the fix is in. ping me on the thread once ok and the build is good.